ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -318,7 +318,12 @@ let compute_max_host_failures_to_tolerate ~__context ?live_set ?protected_vms ()
   let protected_vms = match protected_vms with
     | None -> all_protected_vms ~__context
     | Some vms -> vms in
-  let nhosts = List.length (Db.Host.get_all ~__context) in
+  let total_hosts = List.length (Db.Host.get_all ~__context) in
+  (* For corosync HA less than half of the pool can fail whilst maintaining quorum *)
+  let corosync_ha_max_hosts = ((total_hosts - 1) / 2) in
```

Comment:
I think that for corosync we need to look at loosing only the enabled hosts. If you got 10 hosts, with 5 of them cleanly shutdown we can tolerate the loss of 2 additional hosts, so max failures in that case is 7.
So the failures to tolerate would be: `count_of_disabled_hosts (* these already "failed" *) + ((count_of_enabled_hosts - 1)/2)`.

Otherwise if you want to shutdown hosts (which implies increasing the host failures you tolerate if you're already at the max) you'll be told that you can't shutdown more hosts because you reached the limits of HA failures to tolerate.

Calculating this would need to be done while holding the clustering lock with `Xapi_clustering.with_clustering_lock_if_needed `, and every time we enable or disable a cluster host we will need to update the HA plan. To avoid deadlock the function in xapi_ha_failover.ml won't take any locks, but all its callers will have to (Xapi_cluster_host.enable/disable already does, locks need to be added to other callers).

Reply:
So the check we need is to see if the host is down and clustering is disabled? or can we just check is clustering is disabled?

Reply:
You can check whether the associated Cluster_host is enabled or disabled. There is a helper in `Xapi_clustering.is_clustering_disabled_on_host` that you can use.

Reply:
ok, will need to make some additional updates to the unit test framework to get this to work because currently we don't enable clustering on all hosts